### PR TITLE
fix(voice): start the dispatcher when a session ends with no turns

### PR DESCRIPTION
**Pull request by adityasingh2400**
### Summary

A streamed voice session that produces no audio never terminates. `_dispatch_audio` is the only thing that puts `session_ended` on the queue, and it is only started once a turn has text to synthesize, so a session with zero turns leaves `stream()` waiting on the queue forever.

`_done()` now starts the dispatcher when the session finished without one.

The rest of the change is the lifecycle around that, in `_run_multi_turn`:

- Output emitted by `on_start()` is finalized as part of startup, right after the startup block. It used to stay in an open turn, so a greeting without sentence-final punctuation was held until the session ended or folded into the first user turn.
- The transcription session is closed from a `finally`, and that `finally` runs after the `except` has reported the error. Closing before the report meant a failing `close()` replaced the error the consumer actually needed.
- `_done()` is called only after the producer completes cleanly. An error has already queued its terminal event and a cancelled producer has no consumer left, so neither should start TTS work or wait on it.
- In `stream()`, the graceful producer wait now treats an error as terminal, not just `session_ended`. The producer still has to close the transcription session after reporting an error, so cleanup would otherwise cancel it mid-close.

### Test plan

Six tests in `tests/voice/test_pipeline.py`, all through the public pipeline:

- `..._streamed_audio_input_without_turns` covers the zero-turn hang.
- `..._delivers_on_start_output_during_startup` gates the transcription session on the greeting reaching the consumer.
- `..._on_start_output_is_its_own_turn` pins the same guarantee with a session that yields immediately, so nothing serializes the two.
- `..._error_waits_for_the_session_close_before_cleanup` blocks inside `close()` and asserts the producer is not cancelled, then that cleanup completes, the workflow error reaches the consumer, and no TTS work started.
- `..._failing_close_does_not_replace_the_turn_error` fails the turn and the close, and asserts the consumer gets the turn error.
- `..._cancelled_consumer_closes_the_session_without_further_tts` cancels the consumer and asserts the session still closes with no TTS work.

### CODE REVIEW

**Comment by adityasingh2400** [comment]
The Codex P2 above is correct and it is a gap in this PR rather than something pre-existing. Fixed in bd80b75.

I traced the exact trigger before changing anything. `_add_text` only queues a segment when the splitter returns a complete sentence:

```python
combined_sentences, self._text_buffer = self.tts_settings.text_splitter(self._text_buffer)
if combined_sentences:
    ...
    if self._dispatcher_task is None:
        self._dispatcher_task = asyncio.create_task(self._dispatch_audio())
```

So a greeting from `on_start()` that does not end in sentence punctuation leaves the text in `_text_buffer` with `_dispatcher_task` still `None`. Starting the dispatcher in `_done` then finds `_ordered_tasks` empty and `_completed_session` set, breaks immediately, and emits `session_ended`. The greeting is never handed to TTS and no `turn_ended` is produced.

`_done` now flushes that buffer through `_turn_done` before the terminal event.

One thing worth recording, because my first attempt got it wrong. I initially keyed the check on `self._started_processing_turn or self._text_buffer`, which broke `test_streamed_audio_result_ignores_empty_custom_splitter_chunk`. `_started_processing_turn` is cleared by `_finish_turn`, which the dispatcher calls asynchronously when it sees `turn_ended`, so it is still set at `_done` time for a turn that `_turn_done` already completed, and the turn got finished twice. Keying on `_text_buffer` alone is both narrower and the actual condition, since that is what holds the unsent text.

Verification. `test_streamed_audio_result_finishes_a_pending_turn_on_done` fails on fb43ac48 with `Right contains one more item: 'Hello there'`, which is the greeting never reaching the recording TTS. `tests/voice` is 109 passed with ruff, ruff format and mypy clean.

Three failures remain in `test_voice_workflow_errors_apply_model_and_tool_logging_policies`. I checked those against fb43ac48 before and after, and they fail identically on both, so they are not from this change.

**Comment by adityasingh2400** [comment]
Both directions are right, and the second one is a bug I introduced rather than one I missed. Fixed in e622671.

I had keyed the check on `_text_buffer`, and you are correct that it is not a proxy for an open turn:

- A splitter returning `(text, "")` flushes everything, so the buffer is empty while the turn is still open. `session_ended` went out with no `turn_ended`.
- A workflow that buffers a fragment and then raises reaches `_done` through the pipeline's `finally`, where sending that text starts a TTS request after the error event. The consumer has already stopped at the error, so that is unobservable work against the backend.

There is now explicit state. `_pending_turn` is set in `_start_turn`, cleared in `_turn_done`, and cleared in `_add_error` because nothing after an error is observable. `_done` finalizes on that flag.

That also resolves the reason I avoided `_started_processing_turn` in the first place: it is cleared by `_finish_turn`, which the dispatcher calls asynchronously when it sees `turn_ended`, so it is still set at `_done` time for a turn that was already completed and would finish the turn twice. `_pending_turn` is cleared synchronously by whoever ends the turn, so it does not have that problem.

Two tests, both failing on bd80b751:

- `test_streamed_audio_result_finishes_an_open_turn_with_empty_buffer` uses the `(text, "")` splitter and asserts the full `turn_started, audio, turn_ended, session_ended` sequence.
- `test_streamed_audio_result_does_not_speak_buffered_text_after_an_error` buffers a fragment, raises, and asserts the recording TTS was never called. It fails on the previous commit with `Left contains one more item: 'partial'`.

`tests/voice` is 111 passed, with ruff, ruff format and mypy clean. The three `test_voice_workflow_errors_apply_model_and_tool_logging_policies` failures are unchanged from before this PR, which I checked by running the base commit.

**Comment by adityasingh2400** [comment]
Right again, and I confirmed the mechanism rather than taking it on the badge. Fixed in a638518.

`CancelledError` derives from `BaseException`, so `except Exception` in `_run_multi_turn` does not catch it and `_add_error` never runs, but the `finally` still calls `_done()`. With `_pending_turn` untouched, that finalized the turn and started a TTS request for a consumer that had already gone away.

There is now an explicit `_abandon_turn()` on the result, meaning the turn will never be finalized and `_done` must not start work for it. `_add_error` uses it, and the pipeline calls it from a new `except asyncio.CancelledError` on the outer try, before the existing `except Exception` and before the `finally`.

Scope check worth stating: the single-turn path has no `finally`, so it never reached `_done` on cancellation and is unchanged. The inner `except Exception` around `on_start()` is a separate handler that warns and continues, also unchanged.

`test_streamed_audio_result_abandons_a_pending_turn_on_cancellation` buffers a fragment, does what the cancellation path does, then calls `_done`, and asserts the recording TTS was never invoked. It fails on e622671 with `AttributeError: 'StreamedAudioResult' object has no attribute '_abandon_turn'`.

`tests/voice` is 112 passed, with ruff, ruff format and mypy clean over both changed sources. The three `test_voice_workflow_errors_apply_model_and_tool_logging_policies` failures are unchanged from the base commit.

For what it is worth, this is the third round on this one and each finding has been real, including two where my own fix was the thing at fault. Appreciated.

**Comment by adityasingh2400** [comment]
Correcting myself: I have said three times on this PR that the `test_voice_workflow_errors_apply_model_and_tool_logging_policies` failures were pre-existing and unrelated. That was wrong. They are caused by this PR, and they are fixed in ed51540.

The mistake was in what I compared against. I ran the base commit of my own work, fb43ac48, saw the same three failures, and concluded they predated the change. fb43ac48 is this PR's first commit, not `main`. Running them on `main` at 070b6e15 gives 9 passed. So the breakage arrived with this PR's first commit and I kept reporting it as background noise.

The cause is a race the PR loses. In `stream`, the producer is only awaited gracefully when the terminal event was seen:

```python
if saw_session_end and self.text_generation_task is not None:
    await asyncio.shield(self.text_generation_task)
```

On the error path `stream` breaks on the error event, so `saw_session_end` stays false, and it goes straight to `_cleanup_tasks`, which cancels `text_generation_task` if it is not done. Creating a dispatcher inside `_done` added an await to the producer's `finally`, so the producer was still running at that point and got cancelled. The assertion then failed as `CancelledError` from `.exception()` rather than the expected error. On `main` there was no such await and the producer finished first.

The fix is to not create that dispatcher when the producer failed, since `stream` will never read the `session_ended` it would emit. `_abandon_turn` now also records that, so both the error path and the cancellation path skip it.

`tests/voice` is 115 passed with no failures at all, including the three in question, and the original no-turns test still passes since a clean session does still need the terminal event. ruff, ruff format and mypy are clean.

Worth saying plainly: comparing against the branch point rather than the target branch is what let three real failures sit unexamined across several rounds here.

**Comment by adityasingh2400** [comment]
One note on verifying the above: the `Tests` run for ed51540 came back `action_required` rather than executing, so it needs a maintainer to approve the workflow before CI can confirm it. The three preceding commits ran normally, so this is the fork-approval gate rather than anything about the change.

Until then the evidence is local: `tests/voice` at 115 passed with no failures, including the three `test_voice_workflow_errors_apply_model_and_tool_logging_policies` cases that were red on the previous head.

**Review by seratch — CHANGES_REQUESTED** [review]
Thanks for the contribution. The underlying zero-turn hang is valid, but the current implementation has grown around the wrong lifecycle boundary. `on_start()` output is still not finalized when startup finishes, so a greeting can remain buffered until session shutdown or be combined with the first user response. `_pending_turn` and `_producer_failed` then compensate for that behavior inside `_done()`, coupling clean completion with error and cancellation handling.

Please reset this around producer ownership: finalize any emitted `on_start()` output immediately after the startup block; guarantee STT-session cleanup; call `_done()` only after clean producer completion; and let `_done()` start the dispatcher for a successful session with no dispatcher. Remove the new state flags and cancellation branch. Please cover this through the public pipeline with tests for zero turns, startup output delivered before STT completion, and error/cancellation cleanup without additional TTS work.

**Comment by adityasingh2400** [comment]
Thanks, this was the right call. I had been treating `_done()` as a universal teardown hook and then adding state to tell its callers apart, which is exactly backwards. Rewritten around producer ownership as you described, and the two flags plus the cancellation branch are gone.

`_run_multi_turn` now looks like this:

- The `on_start()` block finalizes its own output. If the workflow emitted anything, `_turn_done()` runs immediately after the startup block, before the transcription session is created. A greeting no longer sits in `_text_buffer` waiting for the first user turn to flush it.
- `create_session()` is followed by its own `try`/`finally` that closes the session. Previously the close shared a `finally` with `_done()`, so a failing close would skip the terminal event and hang the stream. Now the close is unconditional once the session exists, and nothing runs after it in that block.
- `_done()` moved out of `finally` to after the `try`/`except`. Only a clean producer reaches it. The error path has already queued its terminal event, and a cancelled producer has no consumer left to serve, so neither one starts TTS work or waits on it.

That makes the whole `result.py` change five lines: `_done()` starts the dispatcher when the session has none. No `_pending_turn`, no `_producer_failed`, no `_abandon_turn`, and `_add_error` is untouched.

Four tests, all through the public pipeline:

- `test_voicepipeline_streamed_audio_input_without_turns` covers zero turns.
- `test_voicepipeline_delivers_on_start_output_during_startup` gates `transcribe_turns()` on an event the consumer sets when it sees `turn_ended`, so the greeting has to be delivered while the transcription session is still open. If the intro turn were left open the gate never releases.
- `test_voicepipeline_failed_turn_closes_the_session_without_further_tts` and `test_voicepipeline_cancelled_consumer_closes_the_session_without_further_tts` assert the session was closed and that a recording TTS model saw nothing. Both use a splitter that never completes a sentence, so there is buffered text at the moment of failure for a regression to pick up.

On the first two: verified they fail against `main` with the source change reverted, the zero-turn one and the startup one both time out. The other two pass on `main` as well, since `main` never calls `_turn_done()` from `_done()`. They are there to keep `_done()` from drifting back into the `finally`, not to demonstrate the bug.

`make format`, `make lint`, `make typecheck`, `make tests` all pass. Rebased onto current `main`.

**Comment by seratch** [comment]
@adityasingh2400 Can you simplify your comments (e.g., omit the test results etc.)?

File: src/agents/voice/pipeline.py [hunk 1]
```diff
@@ -133,36 +133,46 @@ async def process_turns():
...
                     except Exception as e:
                         log_model_and_tool_action_warning(
                             logger, "Voice workflow on_start failed", e
                         )
 
+                    if emitted_intro:
+                        # Finalize the intro turn as part of startup. Leaving it open would
+                        # hold a greeting with no sentence-final punctuation until the session
+                        # ends, or merge it into the first user turn.
+                        await output._turn_done()
```
[the remark this answers is not in the document]
Comment (adityasingh2400) [reply src/agents/voice/pipeline.py]: I could not reproduce this, and I think the ordering it needs is precluded by `_turn_done()` itself. Detail below, plus a test that removes the objection about the gate.

I probed the exact scenario, an `on_start()` greeting followed by a transcription session that yields its first turn immediately, across twenty schedulings, varying how many chunks the TTS yields (1, 2, 3, 5, 8) and how many times the session yields to the loop before producing (0 through 3). All twenty produced `turn_started, turn_ended, turn_started, turn_ended, session_ended`. Two consecutive user turns with no intro also came out correctly paired.

The reason looks structural. `_turn_done()` ends with `await asyncio.gather(*self._tasks)`, and the segment task carrying `turn_ended` puts it on its local queue before the task completes. That put is what wakes the dispatcher, and the task completion is what resolves the gather, in that order, so the dispatcher is always ahead of `_turn_done()`'s resumption in the ready queue. Once it runs, every local queue is already fully populated, because gather waited for all of them, and `asyncio.Queue.get`/`put` on an unbounded queue do not yield when there is an item or room. So the dispatcher drains through to `turn_ended` and `_finish_turn()` in one slice.

There is also a parity point. Between two user turns the loop calls `_turn_done()` and goes straight back to the `async for`. The intro boundary calls `_turn_done()` and then awaits `create_session()` before the same `async for`. If this race were reachable it would be reachable at the inter-turn boundary on `main` with strictly fewer awaits in between, so it would not be something this change introduces.

The point about the gated test is fair though, so I added `test_voicepipeline_on_start_output_is_its_own_turn`. Same guarantee, but the session yields a turn immediately and nothing serializes the two. It asserts the full sequence `turn_started, audio, turn_ended, turn_started, audio, turn_ended, session_ended` and that TTS saw `["Hello there", "the reply"]` as two separate calls. Verified it fails on `main`, where it collapses to a single turn.

`make format`, `make lint`, `make typecheck` clean. 115 passed in `tests/voice`, 6766 passed in the full suite.

If you can produce an interleaving that breaks this, I will take it, but I would want the reproducer rather than treating it as latent.

**Comment by adityasingh2400** [comment]
Understood, will keep them short from here.

File: src/agents/voice/pipeline.py [hunk 2]
```diff
@@ -133,36 +133,46 @@ async def process_turns():
...
-                            await output._add_text(text_event)
-                        await output._turn_done()
+                    try:
+                        async for input_text in transcription_session.transcribe_turns():
+                            result = self.workflow.run(input_text)
+                            async for text_event in result:
+                                await output._add_text(text_event)
+                            await output._turn_done()
+                    finally:
+                        await transcription_session.close()
```
[the remark this answers is not in the document]
Comment (adityasingh2400) [reply src/agents/voice/pipeline.py]: Good catch, that is a real regression I introduced. Moving the close into an inner `finally` put it before the handler, so a failing `close()` replaced the turn error on the way out and the consumer only saw the cleanup error.

Restructured so the `except` reports first and the `finally` closes after, which also keeps the close unconditional. Added a test with a failing turn and a failing `close()` asserting the consumer gets the turn error, verified failing on the previous commit.

**Review by seratch — CHANGES_REQUESTED** [review]
Thanks for resetting this around producer ownership. The zero-turn and startup behavior now look correct. One lifecycle gap remains: after the producer queues a `VoiceStreamEventError`, it may still be awaiting `transcription_session.close()`, but `stream()` immediately enters cleanup and cancels that producer because graceful producer waiting currently covers only `session_ended`.

Please extend the existing terminal-producer wait to cover the error event while preserving caller cancellation. Add a controlled regression test with a blocking `close()`: verify the producer is not cancelled, release the close, then assert cleanup completes, no additional TTS work starts, and the original workflow error reaches the consumer. Please also update the PR summary to describe the current patch.

**Comment by adityasingh2400** [comment]
Fixed. `stream()` now treats an error as a terminal event for the graceful producer wait, so the close that follows it is no longer cancelled by cleanup.

Added `test_voicepipeline_error_waits_for_the_session_close_before_cleanup`, which blocks inside `close()` and asserts the producer is neither cancelled nor done while the consumer is in its finally, then releases it and checks the workflow error arrives with no TTS work started. It fails on the previous commit.

Also stopped the finalization warning firing on the error path, where the producer re-raises the same exception it queued. PR summary rewritten to match the current patch.

**Review by seratch — APPROVED** [review]
